util/serial/src/lib.rs

```diff
@@ -123,6 +131,64 @@ mod json_u64 {
 #[cfg(feature = "serde_with")]
 pub use json_u64::JsonU64;
 
+#[cfg(feature = "serde_with")]
+mod json_u128 {
+
+    use super::*;
+
+    /// Represents u128 using string, when serializing to Json
+    /// Javascript integers are not 64 bit, and so it is not really proper json.
+    /// Using string avoids issues with some json parsers not handling large
+    /// numbers well.
+    ///
+    /// This does not rely on the serde-json arbitrary precision feature, which
+    /// (we fear) might break other things (e.g. https://github.com/serde-rs/json/issues/505)
+    #[derive(Clone, Copy, Debug, Default, Deserialize, Eq, PartialEq, Hash, Serialize)]
+    #[serde(transparent)]
+    pub struct JsonU128(#[serde(with = "serde_with::rust::display_fromstr")] pub u128);
+
+    impl From<&u128> for JsonU128 {
+        fn from(src: &u128) -> Self {
+            Self(*src)
+        }
+    }
+
+    impl From<&u64> for JsonU128 {
+        fn from(src: &u64) -> Self {
+            Self(*src as u128)
+        }
+    }
+
+    impl From<&JsonU128> for u128 {
+        fn from(src: &JsonU128) -> u128 {
+            src.0
+        }
+    }
+
+    impl From<JsonU128> for u128 {
+        fn from(src: JsonU128) -> u128 {
+            src.0
+        }
+    }
+
+    impl Into<f64> for JsonU128 {
```

Comment:
Is there a particular reason this is needed? It's a little wierd that we use Into to express a lossy conversion

Reply:
We're using this in a flooring / truncating operation for transactions with unnecessarily high precision
From the reserve-auditor repo where its being used:
```rust
    /// Floor integer representation of decimal number to a storable degree of accuracy
    /// Eth transfers often have 18 decimals of accuracy, which is too larget for sql to store by default.
    /// With 9 decimal places of precision, transfers with values up to 100 billion tokens are allowed.
    pub fn floor_value(&self, max_decimals: u32) -> Result<JsonU64, std::num::TryFromIntError> {
        match &self.token_info {
            None => {
                // tokenInfo json field is not guaranteed, and so decimal values cannot be guaranteed
                // Attempt a straight type conversion from JsonU128 to JsonU64 in this case
                return JsonU64::try_from(&self.value);
            }
            Some(token_info) => {
                if token_info.decimals <= max_decimals {
                   return JsonU64::try_from(&self.value);
                }

                let orders_reduced_by = token_info.decimals - max_decimals;
                let denominator = 10_u32.pow(orders_reduced_by) as f64;
                let f_val = f64::from(self.value.clone());
                let truncated_val = (f_val / denominator).trunc() as u64;
                Ok(JsonU64::from(&truncated_val))
            }
        }
    }
```

Reply:
Its necessarily lossy, but that shouldn't matter as we're already reducing the precision by at least one decimal place where its being used

Reply:
Sounds good to me